File: src/lib/sources/arbeitnow.ts
import { fetchJson, findApplyEmail, stripHtml, type JobSource, type SourceJob } from "./types";

interface ArbeitnowJob {
  slug: string;
  company_name: string;
  title: string;
  description: string;
  remote: boolean;
  url: string;
  tags: string[];
  job_types: string[];
  location: string;
  created_at: number;
}

export const arbeitnowSource: JobSource = {
  id: "arbeitnow",
  label: "Arbeitnow",
  requiresNetwork: true,

  async fetch(queries, limit) {
    const payload = await fetchJson<{ data?: ArbeitnowJob[] }>(
      "https://www.arbeitnow.com/api/job-board-api",
    );
    if (!payload?.data) return [];

    const needles = queries.map((query) => query.toLowerCase());
    const matches: SourceJob[] = [];

    for (const job of payload.data) {
      const haystack = `${job.title} ${job.tags?.join(" ") ?? ""}`.toLowerCase();
      if (!needles.some((needle) => haystack.includes(needle) || partialMatch(haystack, needle))) {
        continue;
      }

      const description = stripHtml(job.description ?? "");
      matches.push({
        source: "arbeitnow",
        sourceId: job.slug,
        title: job.title,
        company: job.company_name,
        location: job.location || (job.remote ? "Remote" : ""),
        remote: Boolean(job.remote),
        url: job.url,
        applyEmail: findApplyEmail(description),
        description,
        salaryText: null,
        tags: Array.isArray(job.tags) ? job.tags.slice(0, 12) : [],
        postedAt: job.created_at ? new Date(job.created_at * 1000).toISOString() : null,
      });

      if (matches.length >= limit) break;
    }

    return matches;
  },
};

/** Treats a multi-word query as matched when every significant word appears. */
function partialMatch(haystack: string, needle: string): boolean {
  const words = needle.split(/\s+/).filter((word) => word.length > 3);
  return words.length > 1 && words.every((word) => haystack.includes(word));
}

File: app/sources/arbeitnow.py
from __future__ import annotations

from datetime import datetime, timezone

from app.models import SourceJob
from app.sources.types import fetch_json, find_apply_email, partial_match, strip_html


class ArbeitnowSource:
    id = "arbeitnow"
    label = "Arbeitnow"
    requires_network = True

    def fetch(self, queries: list[str], limit: int) -> list[SourceJob]:
        payload = fetch_json("https://www.arbeitnow.com/api/job-board-api")
        if not payload or not isinstance(payload.get("data"), list):
            return []

        needles = [query.lower() for query in queries]
        matches: list[SourceJob] = []
        for job in payload["data"]:
            haystack = f"{job.get('title') or ''} {' '.join(job.get('tags') or [])}".lower()
            if not any(needle in haystack or partial_match(haystack, needle) for needle in needles):
                continue
            description = strip_html(job.get("description") or "")
            created = job.get("created_at")
            posted_at = (
                datetime.fromtimestamp(created, tz=timezone.utc).isoformat() if created else None
            )
            matches.append(
                SourceJob(
                    source=self.id,
                    source_id=str(job.get("slug") or job.get("url") or job.get("title")),
                    title=job.get("title") or "",
                    company=job.get("company_name") or "Unknown company",
                    location=job.get("location") or ("Remote" if job.get("remote") else ""),
                    remote=bool(job.get("remote")),
                    url=job.get("url") or "",
                    apply_email=find_apply_email(description),
                    description=description,
                    salary_text=None,
                    tags=(job.get("tags") or [])[:12],
                    posted_at=posted_at,
                )
            )
            if len(matches) >= limit:
                break
        return matches
